indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/encoding/Asn1Decoding.kt

```diff
@@ -354,46 +425,53 @@ fun Boolean.Companion.decodeFromAsn1ContentBytes(bytes: ByteArray): Boolean {
  */
 fun String.Companion.decodeFromAsn1ContentBytes(bytes: ByteArray) = bytes.decodeToString()
 
+/**
+ * [Asn1Element.Tag] to the decoded length
+ */
+typealias TagAndLength = Pair<Asn1Element.Tag, Long>
+private val TagAndLength.tag:Asn1Element.Tag get() = first
+private val TagAndLength.length:Long get() = second
 
-private fun ByteIterator.readTlv(): TLV = runRethrowing {
-    if (!hasNext()) throw IllegalArgumentException("Can't read TLV, input empty")
+/**
+ * Reads [TagAndLength] and the number of consumed bytes from the source
+ */
+private fun Source.readTagAndLength(): Pair<TagAndLength, Int> = runRethrowing {
+    if (exhausted()) throw IllegalArgumentException("Can't read TLV, input empty")
 
-    val tag = decodeTag()
+    val tag = readAsn1Tag()
     val length = decodeLength()
-    require(length < 1024 * 1024) { "Heap space" }
-    val value = ByteArray(length) {
-        require(hasNext()) { "Out of bytes to decode" }
-        nextByte()
-    }
-
-    return TLV(Asn1Element.Tag(tag.first,tag.second), value)
+    require(length.first >= 0L) { "Illegal length: $length" }
+    return (tag to length.first) to length.second + tag.encodedTagLength
```

Comment:
these `to` constructions feel too clever by half and just make the code harder to read

use `Pair(...)` please

Reply:
you really maintain `Pair((tag to length.first), (length.second + tag.encodedTagLength)` is more readable?

Reply:
`Pair(TagAndLength(tag, length.first), length.second + tag.encodedTagLength)`, but yes

Reply:
PS if you think it's too much, then split it up into `val tagInfo = ...`, `val parsedLength = ...`, and `Pair(tagInfo, parsedLength)`... i agree it's a mouthful, but it's still a mouthful if you use `to`

Reply:
i don't care enough ;-)